File: fix_jwt.js
const jwt = require('jsonwebtoken');

// Strong secret (must be at least 32 bytes random)
const SECRET_KEY = 'your-very-strong-secret-key-at-least-32-characters-long';

function verifyJwt(token) {
    try {
        // Decode without verification to check algorithm
        const decoded = jwt.decode(token, { complete: true });
        if (!decoded) {
            throw new Error('Invalid token');
        }
        const { header } = decoded;
        // Check for None algorithm
        if (!header.alg || header.alg === 'None' || header.alg === 'none') {
            throw new Error('Algorithm None is not allowed');
        }
        // Verify token with strong secret and algorithm whitelist
        const payload = jwt.verify(token, SECRET_KEY, {
            algorithms: ['HS256'],
            // Additional options: ignoreExpiration? depends on need
        });
        // Validate kid if present
        if (payload.kid) {
            const allowedKids = ['key1', 'key2'];
            if (!allowedKids.includes(payload.kid)) {
                throw new Error('Invalid kid');
            }
        }
        return payload;
    } catch (error) {
        throw new Error(`JWT verification failed: ${error.message}`);
    }
}

module.exports = { verifyJwt };

File: fix_jwt.py
import jwt
from jwt.algorithms import get_default_algorithms

# Strong secret (must be at least 32 bytes random)
SECRET_KEY = "your-very-strong-secret-key-at-least-32-characters-long"

def verify_jwt(token):
    try:
        # Decode without verification first to check algorithm
        unverified = jwt.decode(token, options={"verify_signature": False})
        # Check for None algorithm
        if "alg" not in unverified or unverified.get("alg") == "None":
            raise jwt.InvalidTokenError("Algorithm None is not allowed")
        # Verify token with strong secret
        payload = jwt.decode(
            token,
            SECRET_KEY,
            algorithms=["HS256"],  # Restrict to strong algorithms
            options={"require": ["exp", "iat"]}
        )
        # Validate kid if present (allow only known kids)
        if "kid" in payload:
            allowed_kids = ["key1", "key2"]  # Whitelist
            if payload["kid"] not in allowed_kids:
                raise jwt.InvalidTokenError("Invalid kid")
        return payload
    except jwt.ExpiredSignatureError:
        raise
    except jwt.InvalidTokenError as e:
        raise jwt.InvalidTokenError(f"Invalid token: {e}")
    except Exception as e:
        raise jwt.InvalidTokenError(f"Token verification failed: {e}")
